File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.java
import java.util.*;

class Solution {
  public int solution(int[] topping) {
    int answer = 0;

    HashMap<Integer, Integer> left = new HashMap<>();
    HashMap<Integer, Integer> right = new HashMap<>();

    for(int t : topping) {
      right.put(t, right.getOrDefault(t, 0) + 1);
    }

    for(int t : topping) {
      if(right.get(t) == 1) {
        right.remove(t);
      } else {
        right.put(t, right.get(t) - 1);
      }
      left.put(t, left.getOrDefault(t, 0) + 1);

      if(left.size() == right.size()) {
        answer++;
      }
    }

    return answer;
  }
}

File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.py
def solution(topping):
    answer = 0
    dic1 = {}
    dic2 = {}
    
    for top in topping:
        dic2[top] = dic2.get(top, 0) + 1
    
    for top in topping:
        if dic2[top] == 1:
            del dic2[top]
        else:
            dic2[top] = dic2[top] - 1
        
        dic1[top] = dic1.get(top, 0) + 1
        
        if len(dic1) == len(dic2):
            answer += 1
    
    return answer